File: Workshop3_Stacks + Queues/Warmup_Valid Parentheses_20.cpp
/*
    UCLA ACM ICPC: Interview Track Leetcode Problem Solutions

    Disclaimer: This is not the only valid solution and we do not claim our solutions 
    to be optimal in terms of runtime or memory usage, we are merely giving example 
    solutions to questions for learning purposes only

    Valid Parentheses
    Leetcode Problem 20
    https://leetcode.com/problems/valid-parentheses/
*/

class Solution {
public:

    // The strategy for this problem is to use a character stack.
    // We traverse the string from left to right, putting open
    // parenthesis into the stack, and attemping to match closing
    // parenthesis with the top of the stack.
    bool isValid(string s) {
        stack<char> stack;
        for(int i = 0; i < s.size(); i++) {
            char cur = s[i];

            // If the current character is an open parenthesis,
            // push it onto the stack
            if(cur == '(' || cur == '{' || cur == '[') {
                stack.push(cur);
            }

            // If the current character is ')', first check
            // if the stack is empty. If it is, then ')' can't
            // be matched with anything. Next, check if the 
            // top of the stack is '('. It it isn't, then ')'
            // is matched with the wrong open parenthesis.
            // If it is, then we have a match, and we can
            // remove the open parenthesis from the stack.
            else if(cur == ')') {
                if(stack.empty() || stack.top() != '(')
                    return false;
                else
                    stack.pop();
            }

            // Follow a similar process for '}'
            else if(cur == '}') {
                if(stack.empty() || stack.top() != '{')
                    return false;
                else
                    stack.pop();
            }

            // Follow a similar process for ']'
            else if(cur == ']') {
                if(stack.empty() || stack.top() != '[')
                    return false;
                else
                    stack.pop();
            }
        }
        
        // Finally, we must check if the stack is empty at the end.
        // It is wasn't then we would have extra open parenthesis
        // that aren't matched with any closing parenthesis.
        return stack.empty();
    }
};

File: Workshop3_Stacks + Queues/Warmup_Valid Parentheses_20.py
"""
    UCLA ACM ICPC: Interview Track Leetcode Problem Solutions

    Disclaimer: This is not the only valid solution and we do not claim our solutions 
    to be optimal in terms of runtime or memory usage, we are merely giving example 
    solutions to questions for learning purposes only

    Valid Parentheses
    Leetcode Problem 20
    https://leetcode.com/problems/valid-parentheses/

"""

class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """

        # The strategy for this problem is to traverse through
        # the the string and push open parentheses onto the stack
        # and then pop them off the stack once we enocounter their
        # closed parenthesis counterpart.
        
        # We create a dictionary mapping closed parentheses to 
        # their equivalent open parentheses so we can check if
        # what we popped off of the stack corresponds to the 
        # closed parenthesis we currently have
        par = {")":"(","}":"{","]":"["}
        stack = []
        
        # Traverse through all the characters in s
        # An assumption we make here based on the prompt of
        # the question is that there are only parentheses in
        # this string
        for char in s:
            
            # Since our dictionary has closed parenthesis keys,
            # we know that if the current character is not in
            # the dictionary, it's an open parenthesis
            if char not in par:

                # Add open parentheses to the stack
                stack.append(char)

            # Otherwise, we know it's a closed parenthesis
            else: 

                # If the stack is empty, this means we have a 
                # closed parenthesis without any open parentheses
                # before it, so we know this is invalid
                if len(stack) == 0: return False

                # Otherwise, if the top of the stack constains the 
                # open counterpart to your current closed parenthesis,
                # then you have a valid pair, so you can pop the open
                # parenthesis off the stack
                elif stack[-1] == par[char]:
                    # (arr[-1] in Python gives you the last element in a list, 
                    # which is the top of the stack)
                    stack.pop()
                
                # Otherwise, this is not the correct pairing of open and closed
                # parentheses, so the string has invalid parentheses
                else: return False
        
        # Return True if the stack is empty (all parentheses come in pairs)
        # and False if there are left over open parentheses without a match
        return len(stack) == 0